### DISCUSSION TITLE
Flask teardown request equivalent in Fastapi

### DISCUSSION THREAD

**Discussion by maxwelldorliea** [fastapi/fastapi#6012] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Flask Implementation

from models import storage
.....
.....

@app.teardown_request
def close_session(exception=None):
    storage.close_session()
```

### Description

I am building a rest api with fastapi. I implemented the data layer separately from the fastapi application meaning I do not have direct access to the database session in my fastapi application.

I have access to the storage object which have method like ```close_session``` which allow me to close the current session.

Is there a equivalent of **flask** ```teardown_request``` in **fastapi**?

I have looked at fastapi ```on_event('shutdown')``` and ```on_event('startup')```. These two only runs when the application is shutting down or starting up.

### Operating System

Linux

### FastAPI Version

0.92.0

### Python Version

3.8.10

**Comment by williamjamir** [comment, marked as answer]
It is possible to have a teardown request in FastAPI by using a Dependency. Here is an example:
```python
from fastapi import FastAPI, Depends

app = FastAPI()

async def some_func():
    print('Up')
    try:
        yield
    finally:
        print('Teardown')

@app.get('/')
async def some_router_func(foo=Depends(some_func)):
    print('Some code')

if __name__ == '__main__':
    import uvicorn
    uvicorn.run('x:app')
```

When you run this code and make a GET request to '/', you will see the following output in your console:
```
INFO:     Started server process [14025]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)

Up
Some code
INFO:     127.0.0.1:49389 - "GET / HTTP/1.1" 200 OK
Teardown


Up
Some code
INFO:     127.0.0.1:49393 - "GET / HTTP/1.1" 200 OK
Teardown
```

As you can see, the `some_func` function is executed, and everything before the `yield` is called at the begging of the request.

After the request has been finished (everything from the function was executed), the code after the `yield` is executed (acting exactly as a teardown).

I hope this helps!

**Reply by maxwelldorliea** [reply]
How can I do this for all request without manually doing it for each route?

**Reply by williamjamir** [reply]
You can easily apply a dependency for the whole project using the FastAPI constructor.

```
app = FastAPI(dependencies=[Depends(some_func)])
```

Or segment by router ;)
```python

router = APIRouter(
    dependencies=[Depends(some_func)],
)

@router.get
...
```

Let me know if this answers your question :)

**Reply by maxwelldorliea** [reply]
Thanks

**Reply by crazydevlegend** [reply]
how can i use that dependency in a method?

**Reply by ondiekisteven** [reply]
I think middlewares run after this code. it's not the last code to be run
